### PR TITLE
New: Add UI Localization Framework

### PR DESCRIPTION
**Pull request by Snarkenfaugister**
#### Database Migration
NO

#### Description
Port UI localization framework added to Radarr here: https://github.com/Radarr/Radarr/pull/4577
Would like some feedback on these:
 - What happened with Flemish? Is it safe now?
 - Is it safe to assume `languageProfiles` will always have a list of languages in its first item?
 - `LocalizationServices.cs:L141` Is the deserialization/async correct here? Radarr uses `System.Text.Json` and a `FileStream`.
 - Radarr uses TitleCase for their string names where Sonarr uses camelCase for everything else, is TitleCase preffered for this? (currently sending camelCase)

#### Todos
- [x] Tests
- [ ] Move all UI string literals to language files (planned for separate PR)
- [ ] Set up Weblate for translations (@Qstick has said he can help with this)

#### Issues Fixed or Closed by this PR

*  #749

### CODE REVIEW

File: src/NzbDrone.Core/Localization/Core/en.json
Comment (Taloth) [inline src/NzbDrone.Core/Localization/Core/en.json:4]: ahem...

Comment (Snarkenfaugister) [reply src/NzbDrone.Core/Localization/Core/en.json]: 😐

**Comment by Qstick** [comment]
As a general comment, we chose to implement this backend to allow for HealthChecks, Notifications and Provider Fields to be translated without having two separate (UI and Backend) services consuming the translation assets. It's worked out quite nicely on Radarr side thus far.

**Comment by Snarkenfaugister** [comment]
I didn't realise Radarr has extended this past the UI, makes more sense to have a backend implementation for that.
If there is still a preference for a UI only solution I'll update this, as it looks pretty simple to do.

**Comment by Qstick** [comment]
Well not so much beyond the UI, Health checks and Provider Fields get passed to the UI was my main point, the strings themselves only exist on the back end.

**Comment by Snarkenfaugister** [comment]
@markus101 @Taloth I have moved every UI/HealthCheck string I could find into the translation file in [this branch](https://github.com/Snarkenfaugister/Sonarr/tree/translations).
How would you prefer me to submit it? A separate PR after this is merged, or merged into this one?

**Comment by markus101** [comment]
I think a separate PR would be good, which means we need to review this 😄

File: frontend/src/Settings/UI/UISettings.js
Comment (markus101) [inline frontend/src/Settings/UI/UISettings.js:64]: Left from testing?

**Review by markus101 — CHANGES_REQUESTED** [review]
Left some initial feedback on this, nothing too big, though differs from Radarr a bit.

File: src/NzbDrone.Core/Localization/Core/en.json
Comment (Snarkenfaugister) [reply src/NzbDrone.Core/Localization/Core/en.json:5]: @markus101 While I agree these would initially make for a better dev experience, and would improve code readability initially, I don't think it's sustainable.
If a string gets updated, does the key change too? This would mean updating it in every translation file.
I'm not sure exactly how weblate works, but this would normally mean the text comes up as a new string instead of just picking up changes.
If the key doesn't change, then you're left outdated text, which would just make for a more confusing dev experience.

Comment (markus101) [reply src/NzbDrone.Core/Localization/Core/en.json:5]: > If a string gets updated, does the key change too?

Yes.

> This would mean updating it in every translation file.

Yes, but on the flip side, if you use static keys and you change a translation from `{fileCount} files` to `{fileCount} episode files` every translation is now wrong and unless you audit every translation you simply don't know they're wrong. If you expect French and get English you're going to realize it's not translated, if you expect French and get a different French translation than an English user gets you won't know.

I know there is going to be some churn due to spelling, punctuation, etc, but overall it'll enforce other languages to follow English.

Comment (Snarkenfaugister) [reply src/NzbDrone.Core/Localization/Core/en.json:5]: Fair enough, I can do that.
Other issue is that currently all keys start with a lower case letter due to the way the API generates responses.
Are you ok with that staying the same? e.g. `"browser Reload Required": "Browser Reload Required"`
Or should the API be updated?

Comment (markus101) [reply src/NzbDrone.Core/Localization/Core/en.json:5]: We should be able to override that behaviour, so we can have keys in whatever case they're written in. I'll look at how we can achieve that.

File: frontend/src/Utilities/String/translate.js
Comment (Snarkenfaugister) [reply frontend/src/Utilities/String/translate.js:19]: I have not been able to find a way to get `createAjaxRequest` to work, as the method can't wait for an async response.
@Qstick noticed this was also the only place you didn't use `createAjaxRequest` in Radarr, was it for a similar reason?

File: src/NzbDrone.Core/Localization/Core/en.json
Comment (Snarkenfaugister) [reply src/NzbDrone.Core/Localization/Core/en.json]: Thanks, that would be appreciated.
I have left the keys starting with lower case in the meantime.

**Comment by Snarkenfaugister** [comment]
Addressed above feedback.
Just to note, I dislike using full strings for translation keys and I think it will lead to more confusing down the line than it saves.
I realise this isn't my codebase though, and I won't be the one dealing with it most often, so happy to go with the request.

File: src/NzbDrone.Core/Localization/Core/en.json
Comment (Qstick) [reply src/NzbDrone.Core/Localization/Core/en.json]: > > This would mean updating it in every translation file.
> 
> Yes, but on the flip side, if you use static keys and you change a translation from `{fileCount} files` to `{fileCount} episode files` every translation is now wrong and unless you audit every translation you simply don't know they're wrong. If you expect French and get English you're going to realize it's not translated, if you expect French and get a different French translation than an English user gets you won't know.

As an FYI, Weblate does handle this by marking changes to en as needing review in other languages (for non key changes). Weblate also auto updates the keys based on the master translation so you really only ever have to work in en translation. 

I agree with the use of keys over strings for ease and avoidance of constant double work when messing with strings, but weblate should be able to accommodate the other easily.

File: frontend/src/Utilities/String/translate.js
Comment (markus101) [reply frontend/src/Utilities/String/translate.js:19]: Ooh, `async: false`, that really should be avoided at all costs, delaying loading the app while waiting for the translations is one thing, but blocking the UI thread for the request leads to the UI freezing.

File: src/NzbDrone.Core/Localization/Core/en.json
Comment (markus101) [reply src/NzbDrone.Core/Localization/Core/en.json]: This looks to be our easiest way to achieve it:

```diff
diff --git a/src/Sonarr.Api.V3/Localization/LocalizationResource.cs b/src/Sonarr.Api.V3/Localization/LocalizationResource.cs
index 0be5c7626..7ad219096 100644
--- a/src/Sonarr.Api.V3/Localization/LocalizationResource.cs
+++ b/src/Sonarr.Api.V3/Localization/LocalizationResource.cs
@@ -1,8 +1,41 @@
+using System;
 using System.Collections.Generic;
+using Newtonsoft.Json;
 using Sonarr.Http.REST;
 
 namespace Sonarr.Api.V3.Localization
 {
+    public class LocalizationResourceSerializer : JsonConverter
+    {
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            var resource = value as LocalizationResource;
+            writer.WriteStartObject();
+            writer.WritePropertyName("strings");
+            writer.WriteStartObject();
+
+            foreach (var r in resource.Strings)
+            {
+                writer.WritePropertyName(r.Key);
+                serializer.Serialize(writer, r.Value);
+            }
+
+            writer.WriteEndObject();
+            writer.WriteEndObject();
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override bool CanConvert(Type objectType)
+        {
+            return typeof(LocalizationResource).IsAssignableFrom(objectType);
+        }
+    }
+
+    [JsonConverter(typeof(LocalizationResourceSerializer))]
     public class LocalizationResource : RestResource
     {
         public Dictionary<string, string> Strings { get; set; }
```

Result (the last two added for testing):
```json
{
  "strings": {
    "language": "Language",
    "uiLanguage": "UI Language",
    "uiLanguageHelpText": "Language that Sonarr will use for UI",
    "uiLanguageHelpTextWarning": "Browser Reload Required",
    "Hello World": "Hello World",
    "hello {name}": "hello {name}"
  }
}
```

In the same file to make it easy to test, we could also flatten the object and get rid of the strings, but I'm not too worried about that.

Comment (Snarkenfaugister) [reply src/NzbDrone.Core/Localization/Core/en.json]: Added and flattened. 🙂

File: frontend/src/Utilities/String/translate.js
Comment (Snarkenfaugister) [reply frontend/src/Utilities/String/translate.js:19]: Not sure how to avoid that in this case without somehow reloading the page later or updating translated strings.
I'll have a look into it.

**Comment by Qstick** [comment]
What’s status here? I’m about to hookup Readarr in our weblate instance and thought about it

**Comment by Snarkenfaugister** [comment]
> What’s status here? I’m about to hookup Readarr in our weblate instance and thought about it

I got busy and forgot about it completely.
Markus wants the translation API call the be async, will have another look at it this weekend.

**Comment by Snarkenfaugister** [comment]
I have been unable to find a good solution to support async calls with the current localization approach.
From what I've seen, the best approach would be to separate front and back end localization, maintaining the current method for back-end use but moving to `react-i18n` in the front-end.
Unfortunately I don't have the time to look into it any more, so I'll just close this PR for now.

**Comment by markus101** [comment]
Reopening so it doesn’t get forgotten and someone else can take a look.

**Comment by Snarkenfaugister** [comment]
@markus101 Taking another stab at this.
Recent change is only to `translate.js`, I messed up merging the latest `develop` so just reset the commit history.

This will now load the translations async with the English keys displayed initially, so the UI thread won't be blocked.
The texts will not re-render once the translations have loaded though (if they don't load before the page renders), it will require a move to a different tab then back.
Downside here is that consumers with a slow connection that assume another refresh is required would just be refreshing indefinitely
I think it would be possible to store the localization texts in a way that React can manage the re-render once they're loaded, but I'm still learning it so not sure how that would be done in this project's setup.

Let me know if you're happy with this, otherwise I'll take a look at getting it to re-render later.

**Comment by markus101** [comment]
This will be part of v4.

**Pull request closed without merging**
